File: AlgorithmDataStructure/LeetCode/111-TreeMinDepth/depth.c

#include <stdio.h>

#define INT_MAX 2147483647

struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};


int m_min(int a, int b) { return  a < b ? a : b;}

//递归，从底到上 min(ans)+1
int minDepth(struct TreeNode* root) {
    if (root == NULL) { return 0; }
    if (root->left == NULL && root->right == NULL) { return 1; }
    
    int ans = INT_MAX;
    if (root->left != NULL) {
        ans = m_min(minDepth(root->left), ans);
    }
    if (root->right != NULL) {
        ans = m_min(minDepth(root->right), ans);
    }
    return ans + 1;
}

int main(int argc, const char * argv[]) {
    
    struct TreeNode n11 = {1, NULL, NULL};
    struct TreeNode n12 = {2, NULL, NULL};
    struct TreeNode n1 = {3, &n11, &n12};
    
    struct TreeNode n21 = {1, NULL, NULL};
    struct TreeNode n22 = {2, NULL, NULL};
    struct TreeNode n2 = {3, &n21, &n22};
    
    struct TreeNode n = {0, &n1, &n2};
    
    printf("= %d\n", minDepth(&n));
    
    return 0;
}

File: AlgorithmDataStructure/LeetCode/111-TreeMinDepth/depth.swift


import Foundation

public class TreeNode {
    public var val: Int
    public var left: TreeNode?
    public var right: TreeNode?
    public init(_ val: Int) {
        self.val = val
        self.left = nil
        self.right = nil
    }
}




class Solution {
    
    //广度优先迭代
    func minDepth(_ root: TreeNode?) -> Int {
        if root == nil { return 0 }
        
        var stack = [(TreeNode, Int)]()
        stack.append((root!, 1))
        
        var depth = 0
        while !stack.isEmpty {
            let item = stack.removeFirst()
            let node = item.0
            depth = item.1
            if node.left == nil && node.right == nil { break }
            if node.left != nil {
                stack.append((node.left!, depth + 1))
            }
            if node.right != nil {
                stack.append((node.right!, depth + 1))
            }
        }
        return depth
    }
    
    //深度优先迭代
    func minDepth1(_ root: TreeNode?) -> Int {
        if root == nil { return 0 }
        
        var stack = [(TreeNode, Int)]()
        stack.append((root!, 1))
        
        var depth = Int.max
        while !stack.isEmpty {
            let item = stack.removeLast()
            let node = item.0, currentDepth = item.1
            if node.left == nil && node.right == nil {
                depth = min(currentDepth, depth)
            }
            if node.left != nil {
                stack.append((node.left!, currentDepth + 1))
            }
            if node.right != nil {
                stack.append((node.right!, currentDepth + 1))
            }
        }
        return depth
    }
    
    
}


let n11 = TreeNode(1)
let n12 = TreeNode(2)
var n1 = TreeNode(3)
n1.left = n11
n1.right = n12

let s = Solution()
print(s.minDepth(n1))
